### PR TITLE
gh-113317: Argument Clinic: Add libclinic.clanguage

### PR DESCRIPTION
**Pull request by vstinner**
Add libclinic.lang_c module and move the following classes and functions there:

* CLanguage
* declare_parser()

Add libclinic.codegen and move the following classes there:

* BlockPrinter
* BufferSeries
* Destination

Move these functions to libclinic.function:

* permute_left_option_groups()
* permute_optional_groups()
* permute_right_option_groups()

* Issue: gh-113317

### CODE REVIEW

File: Tools/clinic/libclinic/block_printer.py
Comment (erlend-aasland) [inline Tools/clinic/libclinic/block_printer.py:1]: I'm not sure it makes sense to mix all these global in a `block_printer.py` file. For example the language dict? What about the destinations class? Perhaps there is a better name for this file. "Block printer" does not seem to be a logical name for these globals.

File: Tools/clinic/libclinic/clanguage.py
Comment (erlend-aasland) [inline Tools/clinic/libclinic/clanguage.py:1]: Instead of `clanguage.py` what about `lang_c.py` or `language_c.py`? We also have a PythonLanguage class.

**Comment by vstinner** [comment]
@erlend-aasland: I updated the PR to address your review. Please review the updated PR.

I used git rebase to rewrite the commit message and rename files.

File: Tools/clinic/libclinic/block_printer.py
Comment (vstinner) [reply Tools/clinic/libclinic/block_printer.py:1]: I renamed the file to codegen.py.

File: Tools/clinic/libclinic/clanguage.py
Comment (vstinner) [reply Tools/clinic/libclinic/clanguage.py:1]: I renamed the file to lang_c.py.

**Review by erlend-aasland — APPROVED** [review]
Thanks! I think libclinic.codegen makes sense. Also, it makes sense to put the option group stuff together with the parameter stuff (currently in function.py).

I realise that we already got a libclinic.language.py file with the `PythonLanguage` and `Language` classes. I'm fine with using `clanguage.py` instead of `lang_c.py`, as was your original suggestion. Up to you.

**Comment by vstinner** [comment]
> I realise that we already got a libclinic.language.py file with the PythonLanguage and Language classes. I'm fine with using clanguage.py instead of lang_c.py, as was your original suggestion. Up to you.

I renamed the file to `clanguage.py`.

**Pull request merged**